docs/2019-03-28-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,273 @@
+most-general-unifier (MGU) problem.
+MGU will be detailed in a separate document.
+
+Background
+----------
+
+Recommended reading: [proving One-Path Reachability
+claims](2018-11-08-One-Path-Reachability-Proofs.md).
+
+Similarly to the One-Path Reachability document, we assume all pattern variables
+used in this document to be _extended function-like patterns_, that is patterns
+which can be written as `t ∧ p` where the interpretation of `t` contains at most
+one element and `p` is a predicate.
+
+_Extended constructor patterns_ will be those extended function-like patterns
+for which `t` is a functional term, composed out of constructor-like symbols
+and variables.
+
+
+Definitions
+-----------
+
+### Weak always
+
+Given a formula `φ`, let `[w]φ` denote the formula “weak always” `φ`,
+defined by:
+
+```
+[w]φ := νX.φ ∨ (○X ∧ •⊤)
+```
+
+one consequence of the above is that `[w]φ = φ ∨ (○[w]φ ∧ •⊤)`.
+
+Given this definition of weak always, an all-path reachability claim
+```
+∀x.φ(x) → [w]∃z.ψ(x,z)
+```
+basically states that if `φ(x)` holds for a configuration `γ`, for some `x`,
+then `P(γ)`, where `P(G)` is recursively defined on configurations as:
+* there exists `z` such that `ψ(x,z)` holds for `G`
+* or
+    * `G` is not stuck (`G → • T`) and
+    * `P(G')` for all configurations `G'` in which `G` can transition
+
+
+
+Problem Description
+-------------------
+
+Given a set of all-path reachability claims, of the form `∀x.φ(x) → [w]∃z.ψ(x,z)`,
+we are trying to prove all of them together.
+
+
+Algorithms
+----------
+(detailed in the next section)
+
+
+### Algorithm `proveAllPath`
+
+__Input:__ claims `∀x₁.φ₁ → [w]∃z₁.ψ₁`, `∀x₂.φ₂ → [w]∃z₂.ψ₂`, …, `∀xₙ.φₙ → [w]∃zₙ.ψₙ`
+
+__Output:__ Proved or Unproved
+
+* For each claim `∀x.φ → [w]∃z.ψ`
+    * Let `Goals := { ∀x.φ → [w]∃z.ψ }`
+    * While `Goals` is not empty:
+        * Extract and remove `goal` of the form `∀x.φ → [w]∃z.ψ` from `Goals`
+        * Let `goalᵣₑₘ := ∀x. (φ ∧ ¬∃z.⌈φ ∧ ψ⌉) → [w]∃z.ψ`
+        * If `goalᵣₑₘ` is trivialy valid (i.e., if `φ ∧ ¬∃z.⌈φ ∧ ψ⌉ ≡ ⊥`)
+            * continue to the next goal
+        * If not the first while iteration for this claim:
+            * `(Goals', goal'ᵣₑₘ) := deriveSeq(goalᵣₑₘ, claims)`
+                (alternatively, one can use `(Goals', goal'ᵣₑₘ) := derivePar(goalᵣₑₘ, claims)`)
+        * Else, let `(Goals', goal'ᵣₑₘ) := (∅, goalᵣₑₘ)`
+        * Let `(Goals'', goal''ᵣₑₘ) := derivePar(goal'ᵣₑₘ, axioms)`
+        * If `goal''ᵣₑₘ` is not trivially valid (its lhs is not equivalent to `⊥`)
+            * Return `Unprovable`
+        * Let `Goals := Goals ∪ Goals' ∪ Goals''`
+* Return `Provable`
+
+__Note:__ Since the derivation process can continue indefinitely, one could add
+a bound on the total number of (levels of) expansions attempted before
+returning `Unprovable`.
+
+### Algorithm `derivePar`
+
+__Input:__: goal and set of claims/axioms
+
+* goal: `∀x.φ → [w]∃z.ψ`
+* Either claims or axioms:
```

Comment:
Are claims and axioms going to have the same algorithm? I thought they will have different algorithms, because axioms are about one-path reachability, while claims are all-path reachability.

Reply:
I think it suffices to use `derivePar` for axioms and `deriveSeq` for (all-path reachability) claims.

Reply:
I agree with you that it suffices;  however, using `derivePar` for claims is also sound (although it might cover the same space multiple times), and there were some discussions in the team that it might be easier to implement, so I decided to leave it here, and let a design decision argue which is the best for us to use at the moment.